File: 0039-combination-sum/0039-combination-sum.java
class Solution {
      
    private void findCombinations(int i,int[] arr,int target, List<List<Integer>> ans,List<Integer>ds){
        if(i==arr.length){
            if(target==0){
                ans.add(new ArrayList<>(ds));
            }
        return;
        }
        
        
        if(arr[i]<=target){
            ds.add(arr[i]);
            findCombinations(i,arr,target-arr[i],ans,ds);
            ds.remove(ds.size()-1);
        }
        findCombinations(i+1,arr,target,ans,ds);
        
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans=new ArrayList<>();
        findCombinations(0,candidates,target,ans,new ArrayList<>());
        return ans;
    }
}
    
  

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> all;
        vector<int> partial;
        traverse(all,partial,candidates,0,target);
        return all;
    }
    
private:
    void traverse(vector<vector<int>>& all,
                  vector<int>& partial,
                  vector<int>& candidates,
                  int start,
                 int target){
        if(target==0){
            all.push_back(partial);
            return;
        }
        else{
            for(int j=start;j<candidates.size();j++){
                if(candidates[j]<=target){
                partial.push_back(candidates[j]);
                
                traverse(all,partial,candidates,j,target-candidates[j]);
                partial.pop_back();
    
            }
        }
        }
        
    }
};


      